Vis update

deargle opened the pull request:
I'm opening this so that discussion can happen about is as I go along. I'm still pushing to it. Trying to make commits modular.

deargle commented:
oops wrong commits, facepalm, hold on

deargle commented:
So far, it's just decoupling any jinja rendering from kmapper.py. It's a step towards having kmapper.py be only responsible for building the graph, and letting other files like visuals.py or in future plotlyviz whatever handle all vis. That's this commit: https://github.com/scikit-tda/kepler-mapper/pull/196/commits/6c6b68018d1f7906b33b66518d26856f9bc1fc0c

deargle commented:
commit https://github.com/scikit-tda/kepler-mapper/pull/196/commits/2493f73599754672f7baa9293398172366cdb546 is a _much_ cooler commit. see commit message, but fun things like:

* nodes staying put where you drag them
* focus "locking" on a node?
* the ability to freeze (and unfreeze) all nodes with keystrokes f and x, 
* focus node "pulsing."

deargle commented:
![demo](https://user-images.githubusercontent.com/1174653/91629094-001b5400-e983-11ea-9c1b-972147c626c7.gif)

deargle commented:
@sauln yiss, have a way to save a json config file and load it again. It saves node x and y position, and whether the node is fixed. 

As-is, since this does not save node color as part of the config, this could be used to clumsily test multiple node color values. Generate one vis file for each color mapping, save the config from one of them, and then load that config into all the others. Again, clumsy, and I don't want to do that -- I'd rather have a dropdown to switch between colors.

deargle commented:
Hoo boy, just upgraded D3 from v3 to v6 (not trivial!). One _huge_ benefit of the new d3-force library is that it is [deterministic](https://twitter.com/mbostock/status/725124754701717504?lang=en) (!!) (and apparently has been available since 2016!)

So a given graph will always render the same visually, across browsers and reloads.

https://github.com/scikit-tda/kepler-mapper/pull/196/commits/79ec3bde0431c2f175e87215eea75066e78b2b06

deargle commented:
Dropping python 3.4 because travis was complaining about package pyresistent needing >= 3.5,

![image](https://user-images.githubusercontent.com/1174653/93169627-b51f7180-f6e2-11ea-8813-1403bfea3636.png)

deargle commented:
@sauln can you review so far? I'm largely done with python side, I just enabled passing a 2d array of `color_values` -- separate histograms are calculated for each column. Took some type-casting fenangling in visuals.py, maybe you have a better idea of how to do some of it.

Next I'm going to make the d3 html draw the histograms using d3 instead of a static build at render time. But that's all javascript side at this point.

deargle commented:
Awwww yeah we have functionality for multiple color functions now

![demo](https://user-images.githubusercontent.com/1174653/93391389-4cdca700-f82c-11ea-8209-bc9874591ac5.gif)

sauln commented on kmapper/kmapper.py:
```diff
@@ -735,25 +742,38 @@ def visualize(
+            num_color_value_vectors = color_values.shape[1]
+            num_color_function_names = len(color_function_name)
+            if num_color_value_vectors != num_color_function_names:
+                raise Exception('{} `color_function_names` values found, but {} columns found in color_values. Must be equal.'.format(num_color_function_names, num_color_value_vectors))
```
I noticed the library still says it supports 2.7.  We should probably update this to be >3.6 only.  If we do that, you can use f-strings.

deargle replied:
Cool I'll look into that. And yeah, we say we support some really old python versions. I ❤️ only supporting >=3.6

deargle replied:
haven't done f-string yet. future PR?

sauln commented on kmapper/kmapper.py:
```diff
@@ -735,25 +742,38 @@ def visualize(
-            lens_names = []
+        if color_values is None:
+            if color_function_name is not None:
+                raise Exception('Parameter `color_function_name` was set, while `color_values` was not. Refusing to proceed.')
```
Is it possible to proceed?  What about just having a `warn.warning` saying the colors might not be what you expect and then using the default generated `color_values` below?

deargle replied:
Hmm I can do a more nuanced warn / refusal test here. If color_values is None it'd be okay if color_function_name was set, as long as it was either a string or an array with a single entry. If that were the case, we could match that name up with the default-generated color_values. What can't happen on the html side is where the color_values has a different number of entries than the color_function_name -- I expect a one-to-one mapping when I generate the html, for switching functionality. 

But yeah, they could set color_function_name to override the 'Row number' label that would otherwise be set. I'll change that.

deargle replied:
This logic feels too verbose, but how about this general approach? https://github.com/scikit-tda/kepler-mapper/pull/196/commits/9e487b141b4fd20a30ebfde33582d12ffd0dfafa

sauln replied:
You're right it's not great. not sure if there is an easy solution as there are so many dependent parameters swimming around. 

Seems fine to leave it as is.

deargle replied:
Hmm I'll change `color_function_name` to be a default string argument. Immutable, so okay. Will cut down on logic.

deargle replied:
I tidied up the logic a bit, but I didn't do default string arg

sauln commented on kmapper/visuals.py:
```diff
@@ -184,25 +186,34 @@ def _map_val2color(val, vmin, vmax, colorscale=None):
+    color_values = scaler.fit_transform(color_values)
 
     # "Scaler might have floating point issues, 1.0000...0002". Force max and min
     color_values[color_values > 1] = 1
```
Would `np.clip` be faster?

sauln replied:
:D just noticed this is old code.. no reason to change it then.

sauln commented on kmapper/visuals.py:
```diff
@@ -476,20 +516,69 @@ def _format_tooltip(
+        color_function_names = [color_function_name]
+
+    if color_function_names is None:
+        raise Exception('''`color_function` required but not found in mapper_summary `custom_meta`''')
```
Why would this not be found?  When would a user ever be able to hit this? Could you suggest a fix or place to read more in this exception text?

deargle replied:
visuals.py get really hairy because we have so many public functions, and the earlier one buries color_function_name within custom_meta, but only if it's passed -- here, https://github.com/scikit-tda/kepler-mapper/blob/1fdb48f6902bef6693b0dd54c8421abec26fdac2/kmapper/visuals.py#L218-L244

I'll think some more, but it's a parameter-threading mess right now, yeah.

deargle replied:
What do you think about https://github.com/scikit-tda/kepler-mapper/pull/196/commits/63bdf566d0865563bf81c8aeae2ec6ac00f81b98 ?

sauln replied:
🤕  I hear you and am definitely guilty of not keeping this simpler.  I'd be ok renaming some of these functions with a leading underscore.  Would moving them out of public space make it easier to move forward? I don't believe many of them are in the documentation.

deargle replied:
I'm for moving more out of public. Would need to check that doing so doesn't wreck plotlyviz.py, which relies on importing a lot of those functions rn.

sauln replied:
plotlyviz.py should still be able to use them. Biggest thing is adding a `_` prefix to indicate users should use them directly.

deargle replied:
done

sauln commented on kmapper/visuals.py:
```diff
@@ -476,20 +516,69 @@ def _format_tooltip(
 
 def _node_color_function(member_ids, color_values):
-    return np.mean(color_values[member_ids])
+    return np.mean(color_values[member_ids], axis=0)
```
I would *love* it if the user could supply this function.

deargle replied:
Yeah me too -- I forgot to come back and do this, got distracted with the html/js. We'll have to document it carefully, because with the possibility to pass multiple color_value columns now, the function passed here needs to always do axis=0 to preserve the mapping to the color_value shape. Maybe we... add axis=0 to any function they provide, if it's missing? 🤷 Too opinionated? I think the code would crash if not.

sauln replied:
What about just taking a string argument that is `mean`, `median`, `max`,  or `min` for now?  It's not a complete solution, but should be pretty easy to implement.  I assume that opening up that option will in turn produce requests for different functions also.  When and if that happens we can consider a more general pattern.

deargle replied:
That's a really good idea, will do

deargle replied:
done!

deargle replied:
The way I implemented it, any function available on the `np` base class can be used, specified as a string. I just extract it via `getattr(np, node_function_name)`

deargle replied:
See https://github.com/scikit-tda/kepler-mapper/pull/196/commits/67fcd69ee0fcc70c63fb5e30f9804e6b30d4f552

sauln commented on setup.py:
```diff
@@ -54,7 +54,7 @@
         'Jinja2'
       ],
-      python_requires='>=2.7,!=3.1,!=3.2,!=3.3',
+      python_requires='>=2.7,!=3.1,!=3.2,!=3.3,!=3.4',
```
Let's drop 2.7, and only do `>3.6`.  Does anyone use 3.5 anymore?

deargle replied:
Nope, 3.5 not supported.

sauln commented on setup.py:
```diff
@@ -66,9 +66,9 @@
-        'Programming Language :: Python :: 3.4',
         'Programming Language :: Python :: 3.5',
         'Programming Language :: Python :: 3.6',
+        'Programming Language :: Python :: 3.7',
```
Can add 3.8 here.

sauln replied:
We should also change the testing matrix in the travis and appveyor so these versions get tested.

deargle replied:
I don't see an appveyor.yml file?

sauln replied:
My bad, I think that's only ripser.py that has appveyor setup.

deargle replied:
K i updated

sauln commented on .travis.yml:
```diff
@@ -3,7 +3,6 @@ language: python
 matrix:
   include:
     - python: 2.7
```
Can cut 2.7 and 3.5 from the travis matrix.

deargle replied:
fixed by https://github.com/scikit-tda/kepler-mapper/pull/196/commits/32b94265efa4bba992757bb85c4603fa181931d3

sauln commented on README.md:
```diff
@@ -22,7 +22,7 @@ KeplerMapper can make use of Scikit-Learn API compatible cluster and scaling alg
 
 KeplerMapper requires:
 
-  - Python (>= 2.7 or >= 3.3)
```
`- Python (>= 3.6)`

deargle replied:
fixed by https://github.com/scikit-tda/kepler-mapper/pull/196/commits/32b94265efa4bba992757bb85c4603fa181931d3

sauln commented on README.md:
```diff
@@ -90,4 +90,4 @@ To credit KeplerMapper in your work, please cite both the [JOSS paper](https://d
 
 van Veen et al., (2019). Kepler Mapper: A flexible Python implementation of the Mapper algorithm. Journal of Open Source Software, 4(42), 1315, https://doi.org/10.21105/joss.01315
 
-Hendrik Jacob van Veen, Nathaniel Saul, Eargle, David, & Sam W. Mangham. (2019, October 14). Kepler Mapper: A flexible Python implementation of the Mapper algorithm (Version 1.3.3). Zenodo. http://doi.org/10.5281/zenodo.3485851
```
Just noticed your name here is backwards.. All the other names are `<First name> <Last name>`

deargle replied:
Well shoot, it's like that on zenodo, too

deargle commented:
@sauln That should be everything... I haven't touched docs except in the doc strings, but tbh modifying the docs would be a lot of inertia for me

deargle commented:
@sauln added one more -- the ability to specify multiple node_color_functions for the d3 vis, and toggle among them.
![demo](https://user-images.githubusercontent.com/1174653/94059641-f0373a00-fd9f-11ea-92c5-ec41d1b5d776.gif)

deargle commented:
This is a ridiculous request, but I want to get the hacktoberfest t-shirt, so can we merge this and then I open new pull requests for other vis-related updates? Need to open 4 unique PRs in the month of October to get the shirt :-) https://hacktoberfest.digitalocean.com/

sauln commented:
I'll give this a thorough test run in the next couple of days. Feel free to create a second branch and build off of that and we can merge this more-or-less as is.

The gif looks pretty slick!

deargle commented:
Never mind I'm all set, six pull requests in one day to metasploitable3, hurray for minimum viable product 😄 🚀 

![image](https://user-images.githubusercontent.com/1174653/94971406-34bf8580-04c4-11eb-9cf2-98f7c87e59c6.png)

I have a strange career. topological data libraries one day, security vulnerabilities and firewalls the next

sauln commented:
@deargle Is this still a WIP or is it ready to go?  I'll do a thorough review and test drive once it's ready.  I'd love to get it out and start getting some user feedback soon, I think this is going to be a really well received feature 👍

deargle commented:
It's ready for review!

On Wed, Oct 28, 2020, 7:07 PM Nathaniel Saul <notifications@github.com>
wrote:

> @deargle <https://github.com/deargle> Is this still a WIP or is it ready
> to go? I'll do a thorough review and test drive once it's ready. I'd love
> to get it out and start getting some user feedback soon, I think this is
> going to be a really well received feature 👍
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-tda/kepler-mapper/pull/196#issuecomment-718295617>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAI6Y7KAAFK5RC45QQBPXWTSNC553ANCNFSM4QOYC4AA>
> .
>

sauln commented:
Hey @deargle, apologies for the long delay. 🤷‍♂️ it's been a year. I tried taking this for a test spin, but I couldn't get any of the examples working.  Do the `plot_*.py` scripts in the `examples` directory still work with these changes?  It might be my old environments have gone off.

deargle commented:
@sauln I'll check those scripts, but I had been running certain unit tests one at a time and checking the output `mapper_visualization_output.html`. For example, tests in `test_visuals.py` like `test_visualize_multiple_color_function_and_node_color_functions`, `test_visualize_multiple_node_color_functions`, `test_visualize_multiple_color_functions`, etc. Ran using pytest, e.g.

    pytest test/test_visuals.py::TestVisualHelpers::test_visualize_multiple_color_functions

deargle commented:
I pushed a custom json encoder that handles np arrays, which fixed `plot_horse.py`. `plot_cat.py` and `plot_breast_cancer.py` also appear to be working. 

But `plot_digits.py` is not working because pillow integration was removed in scipy a few dozen versions back, and I'm having trouble getting the custom tooltips to render as nicely as they did before with using pillow directly. I think the issue is that scipy did custom min-max scaling before for pixel values. Hopefully I'll have it soon.

deargle commented:
...and fixed it 6 minutes later :-) the values in the load_digits() dataset range from 0 to 16 inclusive, so I set it to use 5-bit greyscale. Custom tooltips look pretty good, scaling would be a little better, but meh. I didn't push any updates to the examples/output directory though.

deargle commented:
I revisited for no really good reason and fixed it to scale up to 8 bits, making the digit images look just as they did before.

sauln commented:
:phew:  okay, I got the output working from the test so I could verify the functionality works.  The scripts in `examples` run, but don't produce anything with a switcher. 

The final bit is some better docs!

I don't think it would need to much though.

* Could you add multiple color functions to at least a couple of the examples in the examples directory?
* Could you fix the docstrings and add something to the examples in the docstrings?
     *  In particular, I see the docstring is incorrect: `color_values (list or 1d array)` while it should accept a list of 1d arrays for multiple color sets as well. 


In short, functionality is awesome but I'm think it will be reallllly hard for people to discover this functionality. Considering how many people have asked for it, I think we better make it really obvious that it's a major feature of the library.

Once we have some good discoverability / docs, I'm game to bump to 2.0 and send out some announcements on tweeters and various slacks :D

sauln commented:
ps: I pushed a small change to the docs config so they'll render correctly.
pss: if you want some help with docs, let me know! I can free some cycles to help polish them up.

deargle commented:
Great, thanks so much for reviewing it! Yes, I'll take a pass on
documentation, for docstrings and examples. Hopefully tomorrow. I'll ping
when they're ready for you to look at.

On Thu, Jan 14, 2021, 7:07 PM Nathaniel Saul <notifications@github.com>
wrote:

> ps: I pushed a small change to the docs config so they'll render correctly.
> pss: if you want some help with docs, let me know! I can free some cycles
> to help polish them up.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-tda/kepler-mapper/pull/196#issuecomment-760596446>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAI6Y7K3DKWMIRTRRFRY6RTSZ6PPLANCNFSM4QOYC4AA>
> .
>

deargle commented:
I'm not sure where this would go in the README, but I'm dropping a reminder here that the docs need [pandoc](https://pandoc.org/installing.html) to be installed on the system in order to render the jupyter notebooks.

deargle commented:
@sauln See new commit! But wait a tick on pushing, I'm embarrassed by the default styling for the buttons on the html.

deargle commented:
Swanky new button styling
![image](https://user-images.githubusercontent.com/1174653/106065092-03447880-60b8-11eb-9eec-233a308b1e2e.png)

Okay, what further documentation do you think it needs? The new features like the deterministic node positioning aren't in the docs, but maybe they're okay only being in release notes?

deargle commented:
Re-reviewing your most recent comment, I missed the docstring fixes, and only added to one example. I'll address.

deargle commented:
> Could you add multiple color functions to at least a couple of the examples in the examples directory?

I just pushed a commit addressing this

deargle commented:
I'm looking at the readthedocs rn trying to better document all this stuff in there

deargle commented:
I'm going to tack on the ability to enable a search bar to this already-massive PR; with the docs pretty much fixed, I think it will be a short step to make a small addition to the docs to point out the new vis features.

deargle commented:
@sauln if we merge #220 , then documentation for this entire PR is pretty much done. We can go further and add a sklearn "User Guide" style visualization page, but the docstring for visualize makes a nice api page that kind of talks about how to use all of these new features

sauln commented on .travis.yml:
```diff
@@ -0,0 +1,40 @@
+language: python
```
Is this `.travis.yml` used?  I thought I converted all the scikit-tda repos to use github actions now.

deargle replied:
Artifact, I'll remove. This PR had updated .travis.yml before travis integration was removed. So I think the rebase re-added the file.

sauln commented on test/conftest.py:
```diff
@@ -0,0 +1,11 @@
+from kmapper import KeplerMapper
+
+@pytest.fixture
+def sc():
```
what does `sc` stand for?

deargle replied:
predates me, but simplicial complex, I assume. I moved this from a different test file into conftest.py so that all tests could use it.

sauln reviewed (approved):
Looks great! Thanks for doing all the docstrings, examples, and testing! 🙇 

I'm going to go ahead and approve, but I dont think the .travis.yml should be included in the PR.  I assume it was a hiccup in merging git history or something?  The repo should be transitioned off travis now. If that's not the case, let me know and I can finish the migration to github actions.

Besides that, I think some of the files could use a run through the black formatter. Not critical, but it's another thing I've been trying to standardize all the scikit-tda repos on.

Thank you again! I'm stoked to get this new viz out. 

Ohh one last thing.  Could you incrememnt the version to 2.0.0 and add some notes in the release notes?

deargle commented:
All of the below changes apply to kmapper's D3 html visualization.

- upgraded d3 from v3 to v6 -- one huge benefit of the new d3-force library is that it is [deterministic](https://
  twitter.com/mbostock/status/725124754701717504?lang=en), so a given graph will always render the same visually, across browsers and reloads.
- clicking on a node will set it as the "focus node" (the node for which details are shown in the details pane).
  until the user clicks off of the node. That is to say, click-focus is not lost if another node is moused-over.
  Click-focus is released if (1) another node is clicked on, or (2) if the user clicks on the blank canvas.
- hovering over a node will "freeze" it in place until no longer hovering over that node. This makes it easier to
  grab the node.
  If no node is currently set as the "focus node" via a click, then hovering over a node will also make it the focus node.
- once a node is dragged, it stays ("freezes") where it was dragged
- added the ability to freeze (and unfreeze) all nodes with keystrokes f and x,
- the focus node visually "pulses" in the display
- added the ability to "save" the positioning of all nodes in the display. Saves to a .json file.
  Node positioning can be re-loaded via providing the json save file.
- multiple `color_values` arrays can be passed, and switched between interactively in the display.
- the node color function can be specified, as a string, to any function available on the numpy base class (e.g.,
  'mean', 'median', 'max', 'min'. (Before, the only available function was `np.mean`.
    - Multiple node color functions can be specified, and toggled between interactively in the display.
- The toolbar display now uses css flexbox, which avoids overlap-problems on smaller viewports.
- added searchbar functionality, includes AND, OR, EXACT methods (see documentation for `include_searchbar` in `kmapper.visualize()`)

deargle commented:
Whomp, _it is done_

sauln commented on RELEASE.md:
```diff
@@ -26,6 +26,11 @@ All of the below changes apply to kmapper's D3 html visualization.
 
+#### Kmapper
+
+- change several visualize-related functions to be private
```
Could you add version number here like the ones below?

deargle replied:
That's a l4 heading, a subsection of the L3 2.0.0

deargle replied:
We could just remove both of the L4's if you prefer -- the `#### Visualization` and `#### Kmapper`

sauln replied:
ahh yeah sorry I missed that there were multiple parts of the release notes! my bad.

deargle commented:
Oh facepalm, I squashed the black formatting commit into this PR. May the commit history for this massive vis update forever rest in peace.

deargle commented:
Do you prefer to do the PR approval cycle for any and all changes I make to this project, or are you okay with me doing direct commit minutiae for small things, like minor doc tweaks or me removing stray files that I accidentally added to the master branch, like the file kmapper/static/ kmapper-vNext.js ?
